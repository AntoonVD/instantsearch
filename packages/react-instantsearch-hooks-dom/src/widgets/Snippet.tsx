import {
  getHighlightedParts,
  getPropertyByPath,
} from 'instantsearch.js/es/lib/utils';
import React from 'react';

import { Highlight as HighlightUiComponent } from '../ui/Highlight';
import { cx } from '../ui/lib/cx';

import type { ClassNames, PartialKeys } from '../types';
import type { HighlightProps as HighlightUiComponentProps } from '../ui/Highlight';
import type { BaseHit, Hit } from 'instantsearch.js';

export type SnippetProps<THit extends Hit<BaseHit>> = {
  hit: THit;
  attribute: keyof THit | string[];
} & PartialKeys<
  Omit<ClassNames<HighlightUiComponentProps>, 'parts'>,
  'highlightedTagName' | 'nonHighlightedTagName' | 'separator'
>;

export function Snippet<THit extends Hit<BaseHit>>({
  hit,
  attribute,
  classNames = {},
  highlightedTagName,
  nonHighlightedTagName,
  separator,
  ...props
}: SnippetProps<THit>) {
  const property =
    getPropertyByPath(hit._snippetResult, attribute as string) || [];
  const properties = Array.isArray(property) ? property : [property];

  const parts = properties.map((singleValue) =>
    getHighlightedParts(singleValue.value || '')
  );

  return (
    <HighlightUiComponent
      {...props}
      classNames={{
        root: cx('ais-Snippet', classNames.root),
        highlighted: cx('ais-Snippet-highlighted', classNames.highlighted),
        nonHighlighted: cx(
          'ais-Snippet-nonHighlighted',
          classNames.nonHighlighted
        ),
        separator: cx('ais-Snippet-separator', classNames.separator),
      }}
      parts={parts}
      highlightedTagName={highlightedTagName}
      nonHighlightedTagName={nonHighlightedTagName}
      separator={separator}
    />
  );
}
