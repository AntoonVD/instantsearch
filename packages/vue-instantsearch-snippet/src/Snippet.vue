<script>
import { HIGHLIGHT_PRE_TAG, HIGHLIGHT_POST_TAG } from 'instantsearch-store';
import escapeHtml from 'escape-html';

export default {
  functional: true,
  props: {
    result: {
      type: Object,
      required: true,
    },
    attributeName: {
      type: String,
      required: true,
    },
    tagName: {
      type: String,
      default: 'mark',
    },
    escapeHtml: {
      type: Boolean,
      default: true,
    },
  },
  render(h, ctx) {
    const result = ctx.props.result;
    const attributeName = ctx.props.attributeName;
    const tagName = ctx.props.tagName;

    let attributeValue = '';
    if (result._snippetResult && result._snippetResult[attributeName]) {
      attributeValue = result._snippetResult[attributeName].value;
    }

    if (ctx.props.escapeHtml === true) {
      attributeValue = escapeHtml(attributeValue);
    }

    return h('span', {
      class: {
        'ais-snippet': true,
      },
      domProps: {
        innerHTML: attributeValue
          .replace(HIGHLIGHT_PRE_TAG, `<${tagName}>`)
          .replace(HIGHLIGHT_POST_TAG, `</${tagName}>`),
      },
    });
  },
};
</script>
