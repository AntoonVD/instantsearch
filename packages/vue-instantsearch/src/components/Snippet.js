const getProp = (obj, path) =>
  path.split('.').reduce((acc, part) => acc && acc[part], obj);

export default {
  functional: true,
  props: {
    result: {
      type: Object,
      required: true,
    },
    attributeName: {
      type: String,
      required: true,
    },
  },
  render(h, ctx) {
    const result = ctx.props.result;
    const attributeName = ctx.props.attributeName;

    const attributePath = `_snippetResult.${attributeName}.value`;
    const attributeValue = getProp(result, attributePath);

    if (process.env.NODE_ENV !== 'production' && attributeValue === undefined) {
      throw new Error(
        `The "${attributeName}" attribute is currently not configured to be snippeted in Algolia.
        See https://www.algolia.com/doc/api-reference/api-parameters/attributesToSnippet/.`
      );
    }

    return h('span', {
      class: {
        'ais-snippet': true,
      },
      domProps: {
        innerHTML: attributeValue,
      },
    });
  },
};
