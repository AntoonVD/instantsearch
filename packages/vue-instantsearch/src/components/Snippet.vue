<script>import { HIGHLIGHT_PRE_TAG, HIGHLIGHT_POST_TAG } from '../store';
import escapeHtml from 'escape-html';

export default {
  functional: true,
  props: {
    result: {
      type: Object,
      required: true,
    },
    attributeName: {
      type: String,
      required: true,
    },
    tagName: {
      type: String,
      default: 'mark',
    },
    escapeHtml: {
      type: Boolean,
      default: true,
    },
  },
  render(h, ctx) {
    const result = ctx.props.result;
    const attributeName = ctx.props.attributeName;
    const tagName = ctx.props.tagName;

    let attributeValue = '';
    if (result._snippetResult && result._snippetResult[attributeName]) {
      attributeValue = result._snippetResult[attributeName].value;
    }

    if (ctx.props.escapeHtml === true) {
      attributeValue = escapeHtml(attributeValue);
    }

    return h('span', {
      class: {
        'ais-snippet': true,
      },
      domProps: {
        innerHTML: attributeValue
          .split(HIGHLIGHT_PRE_TAG)
          .join(`<${tagName}>`)
          .split(HIGHLIGHT_POST_TAG)
          .join(`</${tagName}>`),
      },
    });
  },
};
</script>
