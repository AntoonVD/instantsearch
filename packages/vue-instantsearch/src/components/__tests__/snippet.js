import Vue from 'vue';
import Snippet from '../Snippet';

test('renders proper HTML', () => {
  const result = {
    _snippetResult: {
      attr: {
        value: `con<em>ten</em>t`,
      },
    },
  };

  const vm = new Vue({
    template: '<snippet attributeName="attr" :result="result">',
    render(h) {
      return h('snippet', {
        props: {
          attributeName: 'attr',
          result,
        },
      });
    },
    components: {
      Snippet,
    },
  }).$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});

test('should render an empty string if attribute is not snippeted', () => {
  const result = {
    _snippetResult: {},
  };

  const vm = new Vue({
    template: '<snippet attributeName="attr" :result="result">',
    render(h) {
      return h('snippet', {
        props: {
          attributeName: 'attr',
          result,
        },
      });
    },
    components: {
      Snippet,
    },
  }).$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});
