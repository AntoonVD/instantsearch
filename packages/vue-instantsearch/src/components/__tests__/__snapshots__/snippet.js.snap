// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders proper HTML 1`] = `

<span class="ais-snippet">
  con
  <em>
    ten
  </em>
  t
</span>

`;

exports[`should render an empty string if attribute is not snippeted 1`] = `

<span class="ais-snippet">
</span>

`;
